### ISSUE TITLE
Questions about mocsy use and packaging

### ISSUE THREAD

**Issue by ocefpaf** [jamesorr/mocsy#2]
xref: https://github.com/ioos/conda-recipes/pull/563#issuecomment-156220667

@jamesorr answer to some of the questions @emiliom and @ocefpaf raised while packaging `mocsy`.

> 1) Yes, mocsy uses EOS-80 and not TEOS-10 because it was the former that was in vigor when the best practices were defined (Dickson et al., 2007). Furthermore, I would not change everything to TEOS10, not yet anyway, because many ocean carbon cycle models still use EOS80. A better option may be to provide functions to convert between TEOS-10 and EOS-80, but this may already be found elsewhere.

That is good to know.  Thanks!  In my simple mind I would just update everything.

> 2) The large subsurface pCO2 differences you are seeing is not an error. See the documentation of the OPTGAS function in the fortran code. Results for pCO2 vary greatly as a function of that option, as detailed in the final publication describing mocsy v2.0 (Fig. 1, section 2.1.3):
> http://www.geosci-model-dev.net/8/485/2015/gmd-8-485-2015.html

In that regard I want to convert the webpage notebooks to unittests to lower the bar for future `mocsy` development.  @jamesorr I will do that in a separate PR ASAP.

> 3) The docstrings in mocsy are those extracted automatically by f2py from the original fortran code. Perhaps they can be improved with options to f2py; I'll need to check.

Don't worry to much about that.  If we go start writing wrapping functions in Python we might want to avoid the automatic generated docstrings.  (The main reason behind this would be to use sphinx to generate docs from the docstrings.)

> Finally, note that 'mocsy' is first and foremost designed for ocean models making carbonate chemistry calculations ONLINE, i.e., during model runs (in Fortran). The OFFLINE calculations, which mocsy allows to be done conveniently in python are a big plus, but I would really like to avoid calling routines in python (e.g., from 'seawater') because that would break the compatibility between the Fortran and python use of mocsy.

That is fine.  (Although I would like to see a comparison of what would break just for the sake if curiosity.)

With that being said we have a decision to make.  I was lazy in the packaging PR #1 and I copied all l the Fortran code into a single file.  Since the design is ONLINE computation that single file is probably not OK.

Solutions:

1) We write a script that generates the single file from the original Fortran code.
2) I do a better job wrapping all the individual files.  (Following more closely what is already done with the [Makefile](https://github.com/jamesorr/mocsy/blob/master/makefile).

(1) Is hackish, but easy.  I am not sure if I can do (2), although that is probably the better way.

I will read more about packing Fortran modules in Python and come up with a better solution for this.  Meanwhile let's avoid merging #1.

**Comment by jamesorr** [comment]
Regarding packing Fortran in Python, I have used f2py for mocsy.  More recently, i have switched to 'f90wrap' (also available on github) for a yet to be released, updated version of mocsy. Are you suggesting to use something else?

The new version mocsy will allow the user to compute the first derivatives of all output variables of the 'vars' routine (e.g., pCO2, pH, CO32-, saturation states) with respect to each input variable (DIC, Alkalinity, temperature, salinity, total phosphorus, total silicon).  These first derivatives are also called sensitivities or buffer factors. In addition, the new version of mocsy will provide error propagation.

**Comment by emiliom** [comment]
@ocefpaf, thanks for your follow ups. I have no experience with f2py, so I have nothing to contribute directly, other than testing to confirm that the docstrings become accessible from Python. Regarding your two options:

> 1) We write a script that generates the single file from the original Fortran code.
> 2) I do a better job wrapping all the individual files. (Following more closely what is already done with the Makefile.

I also have no useful opinion there, since I can't help with option 2.

Regarding your intent to create unit tests, that sounds great. I'm happy to help, but not until the week after this one (I'll be traveling in meetings all week). I think we can plan on keeping the notebooks as additional user help and documentation, and your unit tests, well, for unit/code testing.

@jamesorr, great to hear about the upcoming enhancements to mocsy. Thanks also for your detailed responses. I don't have much to add to what Filipe has said. I'm really glad you liked the notebook. We can refine it later and add it to your github repository when it's finalized.

**Comment by ocefpaf** [comment]
> Regarding packing Fortran in Python, I have used f2py for mocsy. More recently, i have switched to 'f90wrap' (also available on github) for a yet to be released, updated version of mocsy. Are you suggesting to use something else?

I use `f2py` in the #1.  That might be hidden by the  packaging boilerplate I am using, but  when I import `Extension` and `setup` from `numpy.distutils.core`  the Fortran extension will be compiled at install time using `f2py`.

I will take a look how to do this with `f90wrap`.  I never used it before, but it seems to be used to pre-process the Fortran code before passing it to `f2py`.  That might be what we need to avoid the copy-and-paste of the code into 1 source file.

> The new version mocsy will allow the user to compute the first derivatives of all output variables of the 'vars' routine (e.g., pCO2, pH, CO32-, saturation states) with respect to each input variable (DIC, Alkalinity, temperature, salinity, total phosphorus, total silicon). These first derivatives are also called sensitivities or buffer factors. In addition, the new version of mocsy will provide error propagation.

Sounds awesome!  I am not a CO2 guy.  I am here just to help get it packaged, but some people around me are really exited about these additions.

When you say `the new version of mocsy`.  Is that the master branch?  Or should I be targeting this PR to another development branch?

**Comment by jamesorr** [comment]
Regarding the new version of mocsy, I haven't yet put in on github because we are still developing and testing it (with my colleague Jean-Marie Epitalon). It may be another month or two before it is available. es is with dual-number automatic differentiation, which is as accurate as the analytical solution. When it is updated, I'll probably put it on the master branch.

**Comment by ocefpaf** [comment]
Thanks for the info!  That should not affect the efforts to package mocsy.

I just finished reading f90wrap page on GitHub and I do not believe it will help with packing.  It does create the Python module automatically but, as a Python guy, I dislike all those auto-generated classes.

To package with f90wrap one would need to run f90wrap first, then f2py, and create a packaged based on the compiled module and python generated files.  Those steps seems to add a layer of complexity that `numpy.distutils.core` was designed to remove.  (BTW f90wrap uses `numpy.distutils.core` to install itself.)

With that said, from a technical (and packaging biased) point of view, I do not recommend using f90wrap.  Unless mocsy needs some of the unique features of f90wrap.

I believe we have more flexibility to improve the user experience by writing some thin Python wrapper-functions to the Fortran functions.  Or at least just provide a direct access like [this](https://github.com/ocefpaf/mocsy/blob/python_package/mocsy/__init__.py).

Either way, the current form of copying all the Fortran code into 1 file is not acceptable.  I will re-write the `setup.py` to be smarter about this.

**Comment by emiliom** [comment]
> I believe we have more flexibility to improve the user experience by writing some thin Python wrapper-functions to the Fortran functions. Or at least just provide a direct access like this.
> 
> Either way, the current form of copying all the Fortran code into 1 file is not acceptable. I will re-write the setup.py to be smarter about this.

@ocefpaf, let me know how I can help with this. I think I should be able to pitch in a bit. Besides, that will help me learn more about the wrapping and how you create the conda recipes.

But, not until next week, when I'm back home from meetings ...

**Comment by ocefpaf** [comment]
> @ocefpaf, let me know how I can help with this. I think I should be able to pitch in a bit. Besides, that will help me learn more about the wrapping and how you create the conda recipes.

Sure.  I am a little bit swamped these days too.  Lot of work in my day job and I am moving to a new place.  I will be out for a while.

**Comment by emiliom** [comment]
Getting back to mocsy ... @ocefpaf, I [noticed](https://github.com/jamesorr/mocsy/pull/1#issuecomment-161952423) that you [recently looked at options for better fortran wrapping](https://github.com/jamesorr/mocsy/pull/1#issuecomment-161952423). Have you been able to research this further? Earlier you wrote:

> Either way, the current form of copying all the Fortran code into 1 file is not acceptable. I will re-write the setup.py to be smarter about this.

Do you anticipate us being able to make some progress on this in the near future? One thing I'd look forward to (and can offer to help, if you tell me how) is having access from Python to the rich doc strings for each function, as currently found in the Fortran modules. And of course, being able to work directly/automatically off of @jamesorr's master code, to accommodate updates easily.

One small update I can report is that I've edited my [mocsy_examples ipython notebook](https://github.com/emiliom/mocsy/blob/python_package_em/notebooks/mocsy_examples.ipynb) to remove the incorrect statements about suspicious pCO2 results. In `Example 3` I now use `optgas='Pzero'` to match [the output csv file from the mocsy repository](https://github.com/jamesorr/mocsy/blob/master/DIC-Alk-P-vary-from-mocsy.csv), which must have been run with that option rather than the default `optgas='Pinsitu'`. This change also exposes the `optgas` option, which is an important one that wasn't being used in any of the examples.

Thanks. Off to AGU on Monday, for the whole week. But I can work on this while at AGU, if we can coordinate!

**Comment by ocefpaf** [comment]
> Do you anticipate us being able to make some progress on this in the near future? 

It is relatively easy to do.  I just could not find the time.  (I have am important SECOORA deadline for January.)

I will be able to get back to this during the holidays.

**Comment by emiliom** [comment]
Great to hear! Good luck with your SECOORA work. No rush.
